Jetpack Forms: Update the styling on response emails

donnchawp opened the pull request:
Jetpack Forms sends out lots of emails to website owners and they are very basic right now. This PR will add filters to change content more easily and a new, more attractive default style to those emails.

Filters added:
* jetpack_forms_response_email - fired when creating the HTML of the response. An array is passed to it of response field label => value arrays and return a string with formatted HTML if it is to be used.
* jetpack_forms_response_email_title - The title of the email message shown in bold in the email body. Default: "You got a new response!"
* jetpack_forms_response_email_footer - an array of meta information about the response such as time, IP, source URL, and user information.
* jetpack_forms_respone_email_template - fired when wrapping the message in HTML tags. Pass it a string containing the boiler plate open and closing tags of a HTML page with %1$s for the message title, %2$s for the response, %3$s for the response link, %4$s for the link to the form on the website, %5$s for the meta data footer.

This allows website owners to customize the response emails sent from their website.

For context: pcsBup-yo-p2

Emails before:
![image](https://user-images.githubusercontent.com/5120357/236024812-ddceec6f-a788-4ae1-932a-f2853d9f5163.png)

After:

<img src="https://user-images.githubusercontent.com/5120357/236025358-14a5e1e6-244b-4964-b6a9-0a61587e5007.png" width="200px" height="644px">

## Does this pull request change what data or activity we track or use?
No

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pcsBup-yo-p2

## Testing instructions:
Apply patch
Add a contact form to a website and fill it in
Check your inbox.
The email will look much the same as before but you'll see two extra parts:
* A header "You got a new response!" message
* A footer "View Response" message

donnchawp committed:
Beginning of a response email revamp

donnchawp committed:
Copy changes to package too

donnchawp committed:
changelog

donnchawp committed:
Add default email template

donnchawp committed:
Add styling to package too

donnchawp committed:
Update to email styling

donnchawp committed:
Copy changes to package

donnchawp committed:
More updates to the email styling

donnchawp committed:
Fix name of URL field in email message

donnchawp committed:
Remove broken tracking pixel from email

donnchawp committed:
Update version to 0.15

donnchawp committed:
Merge branch 'trunk' into update/jetpack_forms_email_styling

donnchawp committed:
Use a simple check on these arrays

donnchawp committed:
Fix tests for new response emails

donnchawp committed:
Fix more tests

donnchawp committed:
Fix this check in the package too

donnchawp committed:
Update versions and ref hash

donnchawp committed:
Remove this empty line to avoid linting errors

digitalwaveride commented:
@jeherve as Donncha is switching teams next week, could you give it a spin early next week? That would be super awesome, thank you! 🙇

jeherve commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1359,32 +1371,40 @@ public function process_submission() {
+
+		$response_link = admin_url( 'edit.php?post_type=feedback' );
+
+		$message = join( '', apply_filters( 'jetpack_forms_response_email_message', $message ) );
```
Do we need this one, given that we have the existing `contact_form_message` just below? If the intent here is to give more flexibility to folks, maybe we could add the array before it's turned into a string as the second parameter of the existing `contact_form_message` filter?

donnchawp replied:
Good idea. Thanks.

jeherve commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -659,17 +658,30 @@ public static function get_compiled_form_for_email( $feedback_id, $form ) {
 			}
 		}
 
+		$updated_compiled_form = apply_filters( 'jetpack_forms_response_email', $compiled_form, $feedback_id, $form );
```
It would be nice to have docblocks for each one of the new filters, to exactly explain what they can be used for. That documentation is then used to automatically generate [the Jetpack Codex](https://developer.jetpack.com/) on each release. Given [the amount of the filters we already have for the contact form feature](https://developer.jetpack.com/module/contact-form/), it's going to be important that we document the exact use-case for each filter so no one feels overwhelmed or confused as to what filter to use.

You can check the existing filters in these files to get an idea of the expected format of those docblocks. They should all include:
- A title
- `@module contact-form`
- `@since $$next-version$$` (that placeholder is then automatically replaced by the right version number on release)
- a `@param` entry for each parameter.

When documenting filters and the filter is used in multiple files, you can document usage in one place, and only refer to the full documentation in the other. Here is an example:
https://github.com/Automattic/jetpack/blob/0e4a3203030c28c1cb5b6a16b62ccd6a94a1bdf5/projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php#L3741

donnchawp replied:
Thank you for the great explanation. I'll add those.

jeherve commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1603,16 +1627,277 @@ public static function wrap_message_in_html_tags( $body ) {
+					<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
+					<html xmlns="http://www.w3.org/1999/xhtml">
+						<head>
+							<title>Jetpack Forms Response</title>
```
Should this be translatable?

digitalwaveride replied:
👍  Yes, I think so, but only the "Response" part, not the "product name" Jetpack Forms.

donnchawp replied:
I have to make the entire string translatable as some languages will put the word "Responses" in front of "Jetpack Forms". For example, In Irish it would be "Freagraí Jetpack Forms" or perhaps "Jetpack Forms: Freagraí".

jeherve commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1603,16 +1627,277 @@ public static function wrap_message_in_html_tags( $body ) {
+				apply_filters(
+					'jetpack_forms_respone_email_template',
+					'
+					<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
```
Maybe that template could be moved to its own file instead of hardcoding it here, and folks would be able to point to their own template instead?

donnchawp replied:
Done

jeherve commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1586,11 +1606,15 @@ public static function get_mail_content_type() {
 	 * @return string
 	 */
-	public static function wrap_message_in_html_tags( $body ) {
+	public static function wrap_message_in_html_tags( $title, $response_link, $form_link, $body, $footer ) {
```
I think it would be nice to add tests to cover all the changes we're making to that method.

donnchawp replied:
Done!

jeherve commented:
Could you also add some **before** / **after** screenshots in the pull request description? They can provide a quick preview of what the changes entail for anyone finding this PR in the future and looking to understand what the changes were.

Thank you!

digitalwaveride commented:
> Could you also add some **before** / **after** screenshots in the pull request description? 

👍  
Added some screenshots and a "context link" above.

donnchawp committed:
Add doc blocks to filters

donnchawp committed:
Added default email template in separate file.

donnchawp committed:
Load the response template from an external file

donnchawp committed:
Move response email template into a separate file

donnchawp committed:
Set $template first to avoid linter warning and require for tests

Tests run the same code over and over and include_once was only loading
the template once, not each time.

donnchawp committed:
Add tests for wrap_message_in_html_tags()

donnchawp committed:
Merge branch 'trunk' into update/jetpack_forms_email_styling

donnchawp committed:
Fix class name and remove error logging

jeherve commented on projects/plugins/jetpack/modules/contact-form/grunion-response-email-template.php:
```diff
@@ -0,0 +1,277 @@
+
+													<a href="https://www.facebook.com/jetpackme" style="color: #00AADC; text-decoration: none; display: inline-block; padding-right:22px;"><img aria-label="jetpack facebook page" src="https://s0.wp.com/wp-content/mu-plugins/html-emails/themes/jetpack/images/facebook-dark.png" width="20" height="20" style="height: auto; line-height:100%%; outline: none; text-decoration: none; border: 0 no ...
+
+													<a href="https://www.linkedin.com/company/jetpack-for-wordpress/" style="color: #00AADC; text-decoration: none; display: inline-block; padding-right: 22px;"><img aria-label="jetpack linkedin page" src="https://s0.wp.com/wp-content/mu-plugins/html-emails/themes/jetpack/images/linkedin-dark.png" width="20" height="20" style="height: auto; line-height:100%%; outline: none; text-decorati ...
```
I think we could translate `jetpack linkedin page` too, as well as the other labels above.

donnchawp replied:
Done.

jeherve commented on projects/plugins/jetpack/modules/contact-form/grunion-response-email-template.php:
```diff
@@ -0,0 +1,277 @@
+											</tr>
+											<tr>
+												<td align="left" style="border-collapse: collapse;">
+													<p class="footer-november-2022-address-text" style="direction: ltr; color: #101517; font-size: 12px; line-height: 1.5; margin-bottom: 10px; margin-top: 24px;">
```
We have a lot of hardcoded `ltr` and `left` alignments in here. Should we make those dynamic, using `is_rtl` maybe?

donnchawp replied:
Done.

jeherve commented on projects/plugins/jetpack/modules/contact-form/grunion-response-email-template.php:
```diff
@@ -0,0 +1,277 @@
+											<tr>
+												<td align="left" style="border-collapse: collapse;">
+													<p class="footer-november-2022-unsub-text" style="direction: ltr; color: #101517; font-size: 12px; line-height: 1.5;">
+													<b style="font-weight: 600;">' . esc_html__( "Don\'t want these emails?", 'jetpack' ) . '</b>
```
I would recommend using an apostrophe here instead of that escaped single quote (which ends up getting displayed):

<img width="221" alt="image" src="https://user-images.githubusercontent.com/426388/236876731-fc26e862-d1ba-4f07-b0af-b350f0558b3e.png">

```suggestion
													<b style="font-weight: 600;">' . esc_html__( 'Don’t want these emails?', 'jetpack' ) . '</b>
```

jeherve commented on projects/plugins/jetpack/modules/contact-form/grunion-response-email-template.php:
```diff
@@ -0,0 +1,277 @@
+											</tr>
+											<tr>
+												<td align="left" class="footer-november-2022-app-links" style="border-collapse: collapse;">
+													<a href="https://play.google.com/store/apps/details?id=com.jetpack.android" style="color: #00AADC; text-decoration: none; width: 115px; display: inline-block; margin-bottom: 64px;"><img src="https://s0.wp.com/wp-content/mu-plugins/html-emails/themes/jetpack/images/play-store-badge.png" class="footer-november-2022-app-image" width="115" height="33" style="line-height: 100%%;outline: n ...
```
Should we use Jetpack redirects for those links?
PCYsg-pY7-p2

donnchawp replied:
Thanks, just added those.

jeherve commented on projects/plugins/jetpack/modules/contact-form/grunion-response-email-template.php:
```diff
@@ -0,0 +1,277 @@
+												<td align="left" class="footer-november-2022-app-links" style="border-collapse: collapse;">
+													<a href="https://play.google.com/store/apps/details?id=com.jetpack.android" style="color: #00AADC; text-decoration: none; width: 115px; display: inline-block; margin-bottom: 64px;"><img src="https://s0.wp.com/wp-content/mu-plugins/html-emails/themes/jetpack/images/play-store-badge.png" class="footer-november-2022-app-image" width="115" height="33" style="line-height: 100%%;outline: n ...
+
+													<a href="https://apps.apple.com/us/app/jetpack-website-builder/id1565481562" style="color: #00AADC; text-decoration: none; width: 115px; display: inline-block; margin-bottom: 64px;"><img src="https://s0.wp.com/wp-content/mu-plugins/html-emails/themes/jetpack/images/app-store-badge.png" class="footer-november-2022-app-image" width="115" height="33" style="line-height: 100%%; outline:  ...
```
Does loading images from a wp.com subdomain have any positive or negative impact on the deliverability of the emails?
Given that the contact form feature can be used by folks who did not connect their site to WordPress.com, and who did not agree to our ToS, maybe we should load those images from each site instead?

donnchawp replied:
I'm not sure if loading from wp.com helps or not but you're right that they should be served from the site. I've just made that change.

jeherve commented on projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php:
```diff
@@ -3920,32 +3943,56 @@ public function process_submission() {
+			$sent_by_text = sprintf(
+				// translators: the the name of the site.
+				'<br />' . __( 'Sent by a verified %s user.', 'jetpack' ) . '<br />',
+				isset( $GLOBALS['current_site']->site_name ) && $GLOBALS['current_site']->site_name ? $GLOBALS['current_site']->site_name : '"' . get_option( 'blogname' ) . '"'
```
This can probably be simplified. Do we need to use `$GLOBALS['current_site']->site_name`, or can we just rely on `get_option( 'blogname' )`? And can we escape that title?

donnchawp replied:
I think this is to support MU sites, and indeed, looking at the emails I got it gives the "parent" site name. If we're going to change it I'll leave that for another PR.

jeherve commented on projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php:
```diff
@@ -3920,32 +3943,56 @@ public function process_submission() {
+					'<br />',
+					'<hr />',
+					'<span style="font-size: 12px">',
+					__( 'Time:', 'jetpack' ) . ' ' . $time . '<br />',
+					__( 'IP Address:', 'jetpack' ) . ' ' . $comment_author_IP . '<br />', // phpcs:ignore WordPress.NamingConventions.ValidVariableName.VariableNotSnakeCase
+					__( 'Source URL:', 'jetpack' ) . ' ' . $url . '<br />',
```
I would recommend building those earlier, and making the values fully translatable with `sprintf`, so this will work in languages where the value may come before the the field info.

donnchawp replied:
Good idea, done!

jeherve commented on projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php:
```diff
@@ -4147,33 +4195,46 @@ public static function get_mail_content_type() {
+				$template
 			),
-			$body
+			$title,
+			$body,
+			$response_link,
+			$form_link,
```
Should we do some escaping here?

donnchawp replied:
I don't think we can escape $body, $footer or $template easily as they are supposed to have HTML. The two links can be escaped with esc_url() however.

donnchawp replied:
wp_kses with an extensive list of HTML tags?

donnchawp replied:
BTW - the response (in $body) has already been escaped and stripped of HTML tags before it reaches here.

donnchawp replied:
I'll esc_html() the title, esc_url() the two links, but the footer is made up of escaped strings. The body is the same.

jeherve commented on projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php:
```diff
@@ -3955,11 +4002,12 @@ public function process_submission() {
 		 * @since 1.3.1
 		 *
 		 * @param string $message Feedback email message.
+		 * @param string $message Feedback email message as an array
```
Should this be `array` then?

```suggestion
		 * @param array $message Feedback email message as an array
```

jeherve reviewed (changes requested):
I noticed a much more problematic bug: the emails are now sent to the person who submitted the form, instead of sending them to the person who built the form on the site.

donnchawp committed:
Update projects/plugins/jetpack/modules/contact-form/grunion-response-email-template.php

Replace the single quote with an apostrophe

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

donnchawp committed:
Update projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php

Fix typo in comment

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

donnchawp committed:
Translate these labels

donnchawp committed:
Add text direction to template

donnchawp committed:
Fix this "apostrophe"

jeherve commented on projects/packages/forms/src/contact-form/class-contact-form.php:
```diff
@@ -1359,32 +1382,56 @@ public function process_submission() {
 		}
 
-		$message = join( '', $message );
+		$footer = join(
```
Following the conversation in https://github.com/Automattic/jetpack/pull/30517#discussion_r1188387882 , I think we'll want to take this opportunity to move from `join` to `implode` here.

donnchawp replied:
Looks like join is used a few times. I'll replace them all.

donnchawp committed:
Add redirects for mobile app links

donnchawp committed:
Serve images for email from host, for privacy reasons

donnchawp committed:
Adding imgaes to package too

donnchawp committed:
join -> implode

donnchawp committed:
Fix typo in translators comment

donnchawp committed:
Escape these strings

donnchawp committed:
Make footer strings fully translateable.

donnchawp committed:
Use Jetpack_Forms::plugin_url() for image URLs

GRUNION_PLUGIN_URL isn't defined in the package.

donnchawp committed:
Fix the imgae urls, go back to the old way

donnchawp committed:
Use plugins_url() for the image path

donnchawp committed:
Use plugin_dir_url( __DIR__ ) to get the path to the images

donnchawp commented:
I'm not sure why @jeherve saw that the email went to the site owner and the recipient as wp_mail() is called correctly, and hasn't been changed in this PR. It's as if the Reply-To header leaks out and is used to send a second copy to that person.

jeherve committed:
Simplify placeholders and update escaping for URL

jeherve reviewed (commented):
This tests well for me. I updated 19959dfe966d4f7e2d88d6ab043caf8255188e6c to simplify the placeholders we talked about and make things easier to translate for translators. 

I'm wondering if we could improve on the user experience a bit by linking to the post editor when folks want to modify the post. What do you think?

```patch
diff --git a/projects/packages/forms/src/contact-form/class-contact-form.php b/projects/packages/forms/src/contact-form/class-contact-form.php
index 4c1ae67330..1190b0e19f 100644
--- a/projects/packages/forms/src/contact-form/class-contact-form.php
+++ b/projects/packages/forms/src/contact-form/class-contact-form.php
@@ -1300,7 +1300,23 @@ class Contact_Form extends Contact_Form_Shortcode {
 
 		/** This filter is already documented in modules/contact-form/admin.php */
 		$subject = apply_filters( 'contact_form_subject', $contact_form_subject, $all_values );
-		$url     = $block_template || $block_template_part || $widget ? home_url( '/' ) : get_permalink( $post->ID );
+
+		/*
+		 * Links to the feedback and the post.
+		 */
+		if ( $block_template || $block_template_part || $widget ) {
+			$url        = home_url( '/' );
+			$url_editor = home_url( '/' );
+		} else {
+			$url        = get_permalink( $post->ID );
+			$url_editor = add_query_arg(
+				array(
+					'action' => 'edit',
+					'post'   => $post->ID,
+				),
+				admin_url( 'post.php' )
+			);
+		}
 
 		// translators: the time of the form submission.
 		$date_time_format = _x( '%1$s \a\t %2$s', '{$date_format} \a\t {$time_format}', 'jetpack-forms' );
@@ -1462,7 +1478,7 @@ class Contact_Form extends Contact_Form_Shortcode {
 		$message = apply_filters( 'contact_form_message', implode( '', $message ), $message );
 
 		// This is called after `contact_form_message`, in order to preserve back-compat
-		$message = self::wrap_message_in_html_tags( $title, $response_link, $url, $message, $footer );
+		$message = self::wrap_message_in_html_tags( $title, $response_link, $url_editor, $message, $footer );
 
 		update_post_meta( $post_id, '_feedback_email', $this->addslashes_deep( compact( 'to', 'message' ) ) );
 
diff --git a/projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php b/projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php
index d53dbf293d..76295308f8 100644
--- a/projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php
+++ b/projects/plugins/jetpack/modules/contact-form/grunion-contact-form.php
@@ -3861,7 +3861,23 @@ class Grunion_Contact_Form extends Crunion_Contact_Form_Shortcode {
 
 		/** This filter is already documented in modules/contact-form/admin.php */
 		$subject = apply_filters( 'contact_form_subject', $contact_form_subject, $all_values );
-		$url     = $block_template || $block_template_part || $widget ? home_url( '/' ) : get_permalink( $post->ID );
+
+		/*
+		 * Links to the feedback and the post.
+		 */
+		if ( $block_template || $block_template_part || $widget ) {
+			$url        = home_url( '/' );
+			$url_editor = home_url( '/' );
+		} else {
+			$url        = get_permalink( $post->ID );
+			$url_editor = add_query_arg(
+				array(
+					'action' => 'edit',
+					'post'   => $post->ID,
+				),
+				admin_url( 'post.php' )
+			);
+		}
 
 		// translators: the time of the form submission.
 		$date_time_format = _x( '%1$s \a\t %2$s', '{$date_format} \a\t {$time_format}', 'jetpack' );
@@ -4023,7 +4039,7 @@ class Grunion_Contact_Form extends Crunion_Contact_Form_Shortcode {
 		$message = apply_filters( 'contact_form_message', implode( '', $message ), $message );
 
 		// This is called after `contact_form_message`, in order to preserve back-compat
-		$message = self::wrap_message_in_html_tags( $title, $response_link, $url, $message, $footer );
+		$message = self::wrap_message_in_html_tags( $title, $response_link, $url_editor, $message, $footer );
 
 		update_post_meta( $post_id, '_feedback_email', $this->addslashes_deep( compact( 'to', 'message' ) ) );
 

```

jeherve commented:
> I'm not sure why @jeherve saw that the email went to the site owner and the recipient as wp_mail() is called correctly, and hasn't been changed in this PR. It's as if the Reply-To header leaks out and is used to send a second copy to that person.

I haven't been able to reproduce on another site, so I think this may be a fluke with my development environment.  ¯\_(ツ)_/¯

donnchawp committed:
Merge branch 'trunk' into update/jetpack_forms_email_styling

donnchawp commented:
> I'm wondering if we could improve on the user experience a bit by linking to the post editor when folks want to modify the post. What do you think?

Nice! That makes it much easier to do that, thanks! I'll add that in.

donnchawp committed:
Change unsub link to edit post link. Props @jeherve

donnchawp committed:
Merge remote-tracking branch 'origin/trunk' into update/jetpack_forms_email_styling
